fix(insights): Make async query execution use the same cache as sync

Twixes opened the pull request:
## Problem

When a query is executed async (via `/query/`), we aren't able to reuse its results in sync execution (`/insights/` or `/dashboards/`), because we treat `LimitContext.QUERY` and `LimitContext.QUERY_ASYNC` as different in the cache key. This doesn't make sense, because the queries are the same, max query duration is the only difference.

## Changes

This works as a follow-up to #22151. Now, async queries have the same cache key as sync. Also, if cached results are available, we immediately return the `QueryStatus` with results – no polling needed.

## How did you test this code?

Added `test_insight_refreshing_query_async`.

Twixes committed:
Add `test_insight_refreshing_query_async`

It's red.

Twixes committed:
Allow `enqueue_process_query_task` short circuit if cache warm

Twixes committed:
Make cache key consistent between sync and async execution

Twixes committed:
Remove obsolete comment

Twixes committed:
Fix test_execute_async.py

Twixes committed:
Fix `execute_process_query` call

Twixes committed:
Fix `enqueue_process_query_task` call

Twixes committed:
Update expected cache keys

This is to be expected, because now `get_cache_key()` returns the _full_ cache key, i.e. including f"...._{self._limit_context_aliased_for_cache}_v2".

Twixes commented:
@coderabbitai review

Twixes commented:
These comments are pretty dumb honestly

Twixes commented:
@coderabbitai review

Twixes committed:
Update mypy-baseline.txt

aspicer commented on posthog/caching/calculate_results.py:
```diff
@@ -138,8 +138,7 @@ def calculate_for_query_based_insight(
 
-    if "results" not in response:
-        # Translating `CacheMissResponse` to legacy insights shape
+    if response.keys() == {"cache_key"}:  # Translating `CacheMissResponse` to legacy insights shape
```
This feels a little confusing and too knowledgable about the internal structure of CacheMissResponse.

Can we use `isinstance(response, CacheMissResponse)` or something else that doesn't assume internals?

aspicer replied:
Because of pydantic, the code isn't beautiful but it would be like this
```
    try:
        CacheMissResponse(**response)
        return NothingInCacheResult(cache_key=response.get("cache_key"))
    except ValidationError as e:
        return InsightResult(
            # Translating `QueryResponse` to legacy insights shape
            # Only `results` is guaranteed even for non-insight queries, such as `EventsQueryResponse`
            result=response["results"],
            columns=response.get("columns"),
            last_refresh=response.get("last_refresh"),
            cache_key=response.get("cache_key"),
            is_cached=response.get("is_cached", False),
            timezone=response.get("timezone"),
            next_allowed_client_refresh=response.get("next_allowed_client_refresh"),
            timings=response.get("timings"),
        )
```

Twixes replied:
It's annoying that `DatabaseSchemaQuery`, `TimeToSeeDataSessionsQuery`, and `SessionEventsQuerySerializer` responses are dicts, so `process_query()` serializes has to serialize _everything_ to dicts for consistency, causing us to lose out on Pydantic in this layer

Twixes replied:
I like your suggestion in terms of being Pythonic, but I feel like in practice the try:catch: may be even less clear, hence I'm tempted to stick with the dirty `.keys()` check 😅

aspicer replied:
Yeah i agree the try except is ugly and unclear.

I guess at least add a comment in schema.ts in CacheMissResponse that we do this, in case somebody tries to change it and it breaks.

It doesn't feel good though.

Twixes replied:
Got rid of the unclarity around this by refactoring the processing functions to return without dict-ification

aspicer commented on posthog/clickhouse/client/execute_async.py:
```diff
@@ -236,42 +240,66 @@ def kick_off_task(
+    query_status = QueryStatus(id=query_id, team_id=team.id, start_time=datetime.datetime.now(datetime.timezone.utc))
     manager.store_query_status(query_status)
 
+    try:
```
```
    try:
        cached_response = process_query(
            team=team,
            query_json=query_json,
            limit_context=LimitContext.QUERY_ASYNC,
            execution_mode=ExecutionMode.CACHE_ONLY_NEVER_CALCULATE,
        )
        try:
            CacheMissResponse(**cached_response)
        except ValidationError as e:
            query_status.complete = True
            query_status.error = False
            query_status.results = cached_response
            query_status.end_time = datetime.datetime.now(datetime.timezone.utc)
            query_status.expiration_time = query_status.end_time + datetime.timedelta(
                seconds=manager.STATUS_TTL_SECONDS
            )
            manager.store_query_status(query_status)
            return query_status
    except:
        sentry_sdk.capture_exception()  # Carry on async, if we couldn't get to cache
```

Twixes committed:
Make `process_query_...` return actual model instances

This makes `CacheMissResponse` checks easy and obvious.

Twixes committed:
Fix post-refactor issues

Twixes committed:
Stabilize `test_create_two_similarly_named_organizations`

It looks like we're calling `random()` at some point now, which causes the test to see a different disambiguated org slug than before. But this test should not rely on that – `choice()` just needs to be mocked.

Twixes committed:
Also mock in `test_materialized_column_naming`
